### PR TITLE
connector: accept dataSourceId (sId) at creation, migration to backfill

### PR DESCRIPTION
**Pull request by spolu**
## Description

- Accepts and starts sending from front sId at connector creation
- Update DataSourceConfig to carry it
- Migration to backfill

In subsequent PR: enforce its presence, start using it to upsert, nuke dataSourceName

## Risk

Bork connector. Easy to revert. Connector creation and migration tested in dev.

## Deploy Plan

- run SQL migration
- deploy `front` (start sending dataSourceId at creation)
- deploy `connector`
- run ts migration

### CODE REVIEW

File: connectors/src/lib/api/data_source_config.ts [hunk 1]
```diff
@@ -1,6 +1,9 @@
 import type { ConnectorResource } from "@connectors/resources/connector_resource";
 import type { ConnectorModel } from "@connectors/resources/storage/models/connector_model";
-import type { DataSourceConfig } from "@connectors/types/data_source_config";
+import type {
+  DataSourceConfig,
+  DataSourceInfo,
+} from "@connectors/types/data_source_config";
 
 export function dataSourceConfigFromConnector(
```
Comment (flvndvd) [inline connectors/src/lib/api/data_source_config.ts:8]: Would be nice to have those living on the `ConnectorResource`.

File: connectors/src/lib/api/data_source_config.ts [hunk 2]
```diff
@@ -9,6 +12,18 @@ export function dataSourceConfigFromConnector(
...
     workspaceAPIKey: connector.workspaceAPIKey,
     dataSourceName: connector.dataSourceName,
+    dataSourceId: connector.dataSourceId,
+    workspaceId: connector.workspaceId,
+  };
+}
+
+export function dataSourceInfoFromConnector(
+  // TODO(2024-02-14 flav) Remove ConnectorModel once fully bundled in `ConnectorResource`.
+  connector: ConnectorResource | ConnectorModel
```
Comment (flvndvd) [inline connectors/src/lib/api/data_source_config.ts:22]:
```suggestion
  connector: ConnectorResource
```

File: front/migrations/20240904_data_source_id_to_connector.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    const ds = await getDataSource(auth, c.dataSourceName);
+    if (!ds) {
+      logger.error(
+        { workspaceId: c.workspaceId, dataSourceName: c.dataSourceName },
+        "Failed to retrieve data source"
+      );
+      throw new Error("Failed to retrieve data source");
+    }
+
+    if (c.dataSourceName !== ds.name) {
```
Comment (flvndvd) [inline front/migrations/20240904_data_source_id_to_connector.ts:35]: Could that happen?

File: front/migrations/20240904_data_source_id_to_connector.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  }[] = await connectorDB.query(
+    `SELECT "id", "workspaceId", "dataSourceName" FROM connectors WHERE "dataSourceId" IS NULL`,
+    {
+      type: QueryTypes.SELECT,
+    }
+  );
+
+  for (const c of connectors) {
+    const auth = await Authenticator.internalAdminForWorkspace(c.workspaceId);
+    const ds = await getDataSource(auth, c.dataSourceName);
```
Comment (flvndvd) [inline front/migrations/20240904_data_source_id_to_connector.ts:26]: IMO, we should kill this `getDataSource` wrapper, it's one useless layer.

File: front/migrations/20240904_data_source_id_to_connector.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        "Updated connector"
+      );
+    } else {
+      logger.info(
+        {
+          workspaceId: c.workspaceId,
+          connectorId: c.id,
+          dataSourceId: ds.sId,
+          dataSourceName: ds.name,
+          execute,
```
Comment (flvndvd) [inline front/migrations/20240904_data_source_id_to_connector.ts:74]: `nit` execute is already bind to the logger exposed by `makeScript`.

**Review by flvndvd — APPROVED** [review]

File: connectors/src/lib/api/data_source_config.ts (hunk 1 above)
Comment (spolu) [reply connectors/src/lib/api/data_source_config.ts:8]: this is broader refactor

File: connectors/src/lib/api/data_source_config.ts (hunk 2 above)
Comment (spolu) [reply connectors/src/lib/api/data_source_config.ts:22]: this is broader refactor

File: front/migrations/20240904_data_source_id_to_connector.ts (hunk 3 above)
Comment (spolu) [reply front/migrations/20240904_data_source_id_to_connector.ts:35]: don't know. we'll see :)

**Pull request merged**